blocks/datasets/__init__.py

```diff
@@ -0,0 +1,223 @@
+
+        Notes
+        -----
+        This is implemented because simply repeatedly calling :meth:`open`
+        could result in file handlers not being closed, and to increase
+        performance by performing a seek-operation instead of re-opening
+        the file.
+
+        """
+        pass
+
+    @abstractmethod
+    def get_data(self, state=None, request=None, sources=None):
+        """Request data from the dataset.
+
+        Parameters
+        ----------
+        state : object, optional
+            The state as returned by the :meth:`open` method. The dataset
+            can use this to e.g. interact with files when needed.
+        request : object, optional
+            If supported, the request for a particular part of the data
+            e.g. the number of examples to return, or the indices of a
+            particular minimbatch of examples.
+        sources : tuple of strings, optional
+            The data sources to return. By default all data sources are
+            returned.
+
+        .. todo::
+
+           A way for the dataset to communicate which kind of requests it
+           accepts, and a way to communicate what kind of request is being
+           sent when supporting multiple.
+
+        Returns
+        -------
+        tuple
+            A tuple of data.
+
+        """
+        raise NotImplementedError
+
+
+class DataStream(object):
+    """A stream of data.
+
+    A data stream implements the ``__iter__`` protocol, returning a new
+    iterator over the data each time it gets called (in the case of finite
+    data). A single pass over one of these iterators represents an *epoch*.
+
+    The data stream also opens a given dataset and maintains the state
+    (e.g. file handles) if necessary.
+
+    A data stream can provide an iterator over a dataset, but it can also
+    wrap another datastream e.g. to provide caching or some sort of
+    pre-processing.
+
+    Parameters
+    ----------
+    data : :class:`Dataset` or :class:`DataStream`
+        An object that implements the :meth:`open`, :meth:`reset` and
+        :meth:`get_data` methods. Usually this is a :class:`Dataset`
+        instance, but when chaining iterators to e.g. perform caching this
+        can also be a :class:`DataStream` object.
+
+    """
+    def __init__(self, data, iteration_scheme=None, sources=None):
+        data_state = data.open()
+        update_instance(self, locals())
+
+    def __iter__(self):
+        self.data_state = self.data.reset(self.data_state)
+        return DataIterator(self,
+                            iter(self.iteration_scheme)
+                            if self.iteration_scheme else None, self.sources)
+
+    @property
+    def sources(self):
+        if getattr(self, '_sources', None) is None:
+            return self.data.sources
+
+    @sources.setter
+    def sources(self, sources):
+        self._sources = sources
+
+    def open(self):
```

Comment:
But `open` is required even on the base class because it's used when the data stream is wrapped by another data stream. The parent data stream will call `open` on the wrapped data stream, which will propagate down all the data streams till `open` is called on the dataset.